src/mem/alloc.rs

```diff
@@ -0,0 +1,1063 @@
+
+        loop {
+            if val == 0 {
+                break;
+            }
+
+            order += 1;
+            val >>= 1;
+        }
+
+        order
+    }
+}
+
+unsafe impl GlobalAlloc for SvsmAllocator {
+    unsafe fn alloc(&self, layout: Layout) -> *mut u8 {
+        let ret: Result<VirtAddr, ()>;
+        let size = layout.size();
+
+        SLAB_PAGE_SLAB
+            .lock()
+            .adjust_slab_size()
+            .expect("Failed to adjust Slab size");
+
+        if size <= 32 {
+            ret = self.slab_size_32.lock().allocate();
+        } else if size <= 64 {
+            ret = self.slab_size_64.lock().allocate();
+        } else if size <= 128 {
+            ret = self.slab_size_128.lock().allocate();
+        } else if size <= 256 {
+            ret = self.slab_size_256.lock().allocate();
+        } else if size <= 512 {
+            ret = self.slab_size_512.lock().allocate();
+        } else if size <= 1024 {
+            ret = self.slab_size_1024.lock().allocate();
+        } else if size <= 2048 {
+            ret = self.slab_size_2048.lock().allocate();
+        } else if size <= 4096 {
+            ret = allocate_page();
+        } else {
+            let order = SvsmAllocator::get_order(size);
+            if order >= MAX_ORDER {
+                return ptr::null_mut();
+            }
+            ret = allocate_pages(order);
+        }
+
+        if let Err(_e) = ret {
+            return ptr::null_mut();
+        }
+
+        ret.unwrap().as_u64() as *mut u8
+    }
+
+    unsafe fn dealloc(&self, ptr: *mut u8, _layout: Layout) {
+        let virt_addr = VirtAddr::new(ptr as u64);
+
+        let result = ROOT_MEM.lock().get_page_info(virt_addr);
+
+        if let Err(_e) = result {
+            panic!("Freeing unknown memory");
+        }
+
+        let info = result.unwrap();
+
+        match info {
+            Page::Allocated(_ai) => {
+                free_page(virt_addr);
+            }
+            Page::SlabPage(si) => {
+                let slab = si.slab.as_u64() as *mut Slab;
+
+                (*slab).deallocate(virt_addr);
+            }
+            _ => {
+                panic!("Freeing memory on unsupported page type");
+            }
+        }
+    }
+}
+
+#[global_allocator]
+pub static mut ALLOCATOR: SvsmAllocator = SvsmAllocator::new();
+
+#[alloc_error_handler]
+fn alloc_error_handler(layout: alloc::alloc::Layout) -> ! {
+    panic!("Allocation failed: {:?}\n", layout)
+}
+
+fn root_mem_init(pstart: PhysAddr, vstart: VirtAddr, page_count: usize) {
+    {
+        let mut region = ROOT_MEM.lock();
+        region.start_phys = pstart;
+        region.start_virt = vstart;
+        region.page_count = page_count;
+        region.init_memory();
+        // drop lock here so slab initialization does not deadlock
+    }
+
+    if let Err(_e) = SLAB_PAGE_SLAB.lock().init() {
+        panic!("Failed to initialize SLAB_PAGE_SLAB");
+    }
+}
+
+pub fn mem_init() {
```

Comment:
Please, add a documentation comment (three slashes) describing this function a little bit.